File: crates/sl-frontend/src-tauri/src/command/store.rs
use sl_core::environment::LauncherEnv;
use sl_store::{search_store, StoreContentCategory, StoreContentSearchHits, StoreSource};
use tauri::State;
use tokio::sync::RwLock;

#[tauri::command]
pub async fn cmd_search_store(
    query: &str,
    store_src: StoreSource,
    store_category: StoreContentCategory,
    store_page: u32,
    launcher_env_state: State<'_, RwLock<LauncherEnv>>,
) -> Result<StoreContentSearchHits, String> {
    let launcher_env = launcher_env_state.read().await;
    let requester = launcher_env.requester();
    let result = search_store(
        &requester, 
        query, 
        store_category, 
        store_page, 
        store_src
    ).await
    .map_err(|e| e.to_string())?;
    Ok(result)
}

// #[tauri::command]
// pub async fn fetch_content_versions(
//     store_type: StoreType,
//     slug: &str,
//     game_version: Option<&str>,
//     loader: Option<&str>,
//     launcher_env_state: State<'_, RwLock<LauncherEnv>>,
// ) -> Result<StoreProjectVersions, String> {
//     let launcher_env = launcher_env_state.read().await;
//     let requester = launcher_env.requester();

//     let result = get_content_versions(store_type, slug, game_version, loader, requester)
//         .await
//         .map_err(|e| e.to_string())?;

//     Ok(result)
// }

// #[tauri::command]
// pub async fn install_content(
//     instance_name: &str,
//     files: Vec<ContentFile>,
//     launcher_env_state: State<'_, RwLock<LauncherEnv>>,
// ) -> Result<(), String> {
//     let launcher_env = launcher_env_state.read().await;

//     download_content(
//         &launcher_env.requester(),
//         &launcher_env.instances(),
//         instance_name,
//         files,
//     )
//     .await
//     .map_err(|e| e.to_string())?;

//     Ok(())
// }

// #[tauri::command]
// pub async fn install_modpack(
//     slug: &str,
//     version: &str,
//     launcher_env_state: State<'_, RwLock<LauncherEnv>>,
// ) -> Result<(), String> {
//     let env = launcher_env_state.write().await;
//     download_modrinth_modpack(&mut env.instances(), slug, version)
//         .await
//         .map_err(|e| e.to_string())?;

//     Ok(())
// }

File: crates/sl-frontend/src/lib/managers/store.ts
import { reactive } from "vue";
import { invoke } from "@tauri-apps/api/core";
import { StoreContentCategory, StoreContentHit, StoreContentHits, StoreContentVersion, StoreSource } from "@/types/store";
import { InstanceMetadata } from "@/types/instances";
import { setIfNotPresent } from "../utils";

export interface SelectedContentState {
    slug: string,
    versions: StoreContentVersion[],
    loading: boolean
}

export const storeManager = reactive({
    storeType: 'Modrinth' as StoreSource,
    storeCategory: 'modpacks' as StoreContentCategory,
    searchQuery: "",
    page: 1,

    items: undefined as StoreContentHits | undefined,
    loading: false,

    selectedContent: reactive<SelectedContentState>({
        slug: "",
        versions: [],
        loading: false
    }),
    selectedContents: new Map<string, StoreContentVersion | undefined>,
    selectedInstance: undefined as InstanceMetadata | undefined,

    selectContent: async (slug: string) => {
        if (storeManager.storeCategory === 'modpacks') {
            storeManager.selectedContents.clear()
            storeManager.selectedContent.slug = slug
            setIfNotPresent(storeManager.selectedContents, slug, undefined)
        } else {
            if (storeManager.selectedContents.has(slug) && storeManager.selectedContent.slug == slug) {
                storeManager.selectedContents.delete(slug)
                storeManager.selectedContent.slug = ""
            } else {
                storeManager.selectedContent.slug = slug
                setIfNotPresent(storeManager.selectedContents, slug, undefined)
            }
        }
    },

    fetchSearch: async (
        query: string,
        storeSrc: StoreSource,
        storeCategory: StoreContentCategory,
        storePage: number
    ) => {
        const result = await invoke<StoreContentHits>("cmd_search_store", {
            query: query,
            storeSrc: storeSrc,
            storeCategory: storeCategory,
            storePage: (storePage - 1)
        });

        return result
    },

    loadSearch: async () => {
        storeManager.loading = true
        try {
            storeManager.items = await storeManager.fetchSearch(
                storeManager.searchQuery,
                storeManager.storeType,
                storeManager.storeCategory,
                storeManager.page
            )
        } catch (e) {
            console.error("storeManager.loadSearch error: ", e);
        } finally {
            storeManager.loading = false
        }
    },

    // fetchContentVersions: async () => {
    //     if (storeManager.storeCategory === 'modpacks') {
    //         const result = await invoke<StoreContentVersion[]>("fetch_content_versions", {
    //             storeType: storeManager.storeType,
    //             slug: storeManager.selectedContent.slug,
    //             gameVersion: undefined,
    //             loader: undefined
    //         })
    //         return result
    //     }

    //     const result = await invoke<StoreContentVersion[]>("fetch_content_versions", {
    //         storeType: storeManager.storeType,
    //         slug: storeManager.selectedContent.slug,
    //         gameVersion: storeManager.selectedInstance?.mc_version,
    //         loader: storeManager.selectedInstance?.mod_loader
    //     })

    //     console.log(result)

    //     return result
    // },

    // loadContentVersions: async () => {
    //     storeManager.selectedContent.loading = true;
    //     try {
    //         storeManager.selectedContent.versions = await storeManager.fetchContentVersions()
    //         setIfNotPresent(storeManager.selectedContents, storeManager.selectedContent.slug, storeManager.selectedContent.versions[0])
    //     } catch (error) {
    //     } finally {
    //         storeManager.selectedContent.loading = false;
    //     }
    // },

    // installContent: async () => {
    //     try {
    //         await invoke("install_modpack", {
    //             slug: storeManager.selectedContent.slug,
    //             version: storeManager.selectedContents.get(storeManager.selectedContent.slug)?.id
    //         })
    //         // await invoke("install_content", {
    //         //     instanceName: storeManager.selectedInstance?.name,
    //         //     files: Array.from(storeManager.selectedContents.values()).map(val => val?.files[0])
    //         // })
    //         console.log("StoreManager.installContent log: Successfully installed")
    //     } catch (error) {
    //         console.error("StoreManager.installContent error: ", error)
    //     }
    // },
})